sdk/identity/azure-identity/src/private/chained_token_credential_impl.hpp

```diff
@@ -5,21 +5,36 @@
 
 #include "azure/identity/chained_token_credential.hpp"
 
+#if defined(TESTING_BUILD)
+class DefaultAzureCredential_CachingCredential_Test;
+#endif
+
 namespace Azure { namespace Identity { namespace _detail {
 
   class ChainedTokenCredentialImpl final {
+
+#if defined(TESTING_BUILD)
+    // make tests classes friends to validate caching
+    friend class DefaultAzureCredential_CachingCredential_Test;
+#endif
+
   public:
     ChainedTokenCredentialImpl(
         std::string const& credentialName,
-        ChainedTokenCredential::Sources&& sources);
+        ChainedTokenCredential::Sources&& sources,
+        bool cacheSelectedCredential = false);
 
     Core::Credentials::AccessToken GetToken(
         std::string const& credentialName,
         Core::Credentials::TokenRequestContext const& tokenRequestContext,
-        Core::Context const& context) const;
+        Core::Context const& context);
 
   private:
-    ChainedTokenCredential::Sources m_sources;
+    std::vector<std::shared_ptr<Core::Credentials::TokenCredential>> m_sources;
+    bool m_cacheSelectedCredential;
+    // This needs to be atomic so that sentinel comparison is thread safe.
+    std::atomic<std::size_t> m_SelectedCredentialIndex = std::numeric_limits<std::size_t>::max();
+    std::mutex m_cachingMutex;
   };
```

Comment:
For better alignment in memory, you can arrange the members as such:
```cpp
ChainedTokenCredential::Sources
std::mutex
(constexpr std::size_t)
std::atomic<std::size_t>
bool
```

Reply:
If we think memory alignment of our objects is important in C++, we should add our policy to the C++ SDK guidelines under the implementation section, and drive consistency across the board.

I don't think we have been particularly focused on that, so far.

Reply:
#5145